tools/nl/embeddings/build_embeddings.py

```diff
@@ -0,0 +1,187 @@
+
+  text2sv[name] = svs_list
+
+
+def get_texts_dcids(df):
+  text2sv_dict = {}
+  for _, row in df.iterrows():
+    name = row[COL_NAME].strip()
+    sv = row[COL_ID].strip()
+    add_sv(name, sv, text2sv_dict)
+    if COL_DESCRIPTION in row:
+      for desc in row[COL_DESCRIPTION].split(';'):
+        descrip = desc.strip()
+        add_sv(descrip, sv, text2sv_dict)
+    if COL_ALTERNATIVES in row:
+      alternatives = row[COL_ALTERNATIVES].split(';')
+      # if len(alternatives) > 1:
+      #   # TODO: check why the last sentence alternative is usually truncated.
+      #   # Ignoring the last alternative because typically that is an incomplete sentence.
+      #   alternatives = alternatives[:-1]
+
+      for palm_alt in alternatives:
+        alt = palm_alt.strip()
+        add_sv(alt, sv, text2sv_dict)
+
+  texts = sorted(list(text2sv_dict.keys()))
+  dcids = [','.join(text2sv_dict[k]) for k in texts]
+
+  return (texts, dcids)
+
+
+def get_sheets_data(ctx, sheets_url, worksheet_name):
+  sheet = ctx.gs.open_by_url(sheets_url).worksheet(worksheet_name)
+  df = pd.DataFrame(sheet.get_all_records())
+  return df
+
+
+def trim_columns(df):
+  cols = [COL_ID, COL_NAME]
+  if COL_DESCRIPTION in df.columns:
+    cols.append(COL_DESCRIPTION)
+  if COL_ALTERNATIVES in df.columns:
+    cols.append(COL_ALTERNATIVES)
+  return df[cols]
+
+
+def two_digits(number):
+  number_str = str(number)
+  if len(number_str) == 1:
+    number_str = "0" + number_str
+  return number_str
+
+
+def build_embeddings(ctx, texts, dcids):
+  embeddings = ctx.model.encode(texts, show_progress_bar=True)
+  embeddings = pd.DataFrame(embeddings)
+  embeddings['dcid'] = dcids
+  embeddings['sentence'] = texts
+  return embeddings
+
+
+def build(ctx):
+  now = datetime.datetime.now()
+  for cf in CSV_FILES:
+    csv_name = cf["name"]
+
+    month_str = two_digits(now.month)
+    day_str = two_digits(now.day)
+    hour_str = two_digits(now.hour)
+    minute_str = two_digits(now.minute)
+    second_str = two_digits(now.second)
+
+    csv_out_name_date = f"{csv_name}_{now.year}_{month_str}_{day_str}_{hour_str}_{minute_str}_{second_str}.csv"
+    csv_out_path = os.path.join(ctx.tmp, csv_out_name_date)
+
+    print("Processing: " + csv_name)
+    print("\t\t getting sheet:", cf["url"])
+
+    df = get_sheets_data(ctx, cf["url"], cf["worksheet"])
+    df = df.fillna("")
+
+    print("Updating input CSV file")
+    df.to_csv(os.path.join('sheets', csv_name + '.csv'), index=False)
+
+    print("Getting texts, dcids and embeddings.")
+    df = trim_columns(df)
+    (texts, dcids) = get_texts_dcids(df)
+
+    print("Building embeddings")
+    embeddings = build_embeddings(ctx, texts, dcids)
+
+    print("Saving locally to " + csv_out_path)
+    embeddings.to_csv(csv_out_path, index=False)
+
+    # Also upload to the NL embeddings server's GCS bucket
+    print("Attempting to write to GCS bucket:")
+    print(f"\t file: gs://{FLAGS.bucket_name}/{csv_out_name_date}")
```

Comment:
nit: since this is the last line that gets printed and the README asks to capture the version number, could we add some text that explicitly says something like:

"Embeddings Filename: <filename>"

And this <filename> should just be {csv_out_name_date} and not the full gcs path.

Reply:
Good call, done.